### PR TITLE
Split the `Layer` protocol into `Layer` and `Module` where `Module` does not require `Input: Differentiable`.

### PR DESCRIPTION
**Pull request by eaplatanios**
`Layer` currently requires `Input: Differentiable`. This results in boilerplate code for layers with non-differentiable inputs (e.g., embedding layers that take integer tensors as input). However, the constraint is required for chaining together layers (e.g., as done in the `Sequential` layer). This PR includes the following changes:

- Introduced a new `Module` protocol that is equivalent to the current `Layer` with the `Input: Differentiable` constraint removed and `callAsFunction` annotated with `@differentiable(wrt: self)` as opposed to `@differentiable(wrt: (self, input))`.
- Conformed `Layer` to `Module` with the additional `@differentiable(wrt: (self, input))` attribute on `callAsFunction`.
- Switched `Embedding` from conforming to `Layer` to conforming to `Module` and removed the boilerplate `EmbeddingInput` struct.
- Adapted `Sequential` so that it can work with both `Module`s and `Layer`s, with the constraint that `Module`s can only appear in the beginning of a sequential layer chain.

### CODE REVIEW

**Comment by rxwei** [comment]
> 1. Rename `XXX` to something better :) I don't like `Module` because modules are usually used to refer to what we call `Layer`, which would cause confusion. Do you have any other ideas? I thought of `InputLayer` but that is too restrictive a description semantically.

`InputLayer` would make models look weird. `Model` would make the `Embedding` layer look weird. Maybe there's something in between?

**Comment by rxwei** [comment]
It seems to be a bug. Do you wanna fix it? :)
You can start by pulling one of those parameterless layers out to a single file, compile it with `-Xllvm -debug-only=differentiation`, and look at the activity analysis result.

**Comment by Shashi456** [comment]
Can we simply name them LayerV1, LayerV2?

**Comment by rxwei** [comment]
> Can we simply name them LayerV1, LayerV2?

We cannot. These are not user-friendly names but merely a reflection of implementation details that users do not need to know.

**Comment by Shashi456** [comment]
Sorry for bouncing name ideas, but `FeatureLayer` ? Initial layers since most often embedding are often feature representations of the input. And even in sequential, the name would seem neat.

**Comment by eaplatanios** [comment]
@rxwei this is ready for review. All tests pass now. I went with `Module` after all, as you initially suggested. I couldn't come up with something better and I think this is not a good reason to delay this further. I haven't been able to debug the invalid `result does not depend on differentiation arguments` warnings but have verified that all tests pass and thus the warnings have no effect on the derivatives. I believe activity analysis may be being triggered twice, once for `Module` and once for `Layer`, due to the overridden function declaration. The warning is probably then triggered by the `Module` function analysis, but the `Layer` function analysis still goes through and results in gradients being propagated correctly.

In either case, since tests pass and this removes an annoying restriction I believe we can go ahead review and merge this.

**Comment by eaplatanios** [comment]
Has the CI toolchain been updated? Because I believe the errors are due to an older toolchain.

**Comment by rxwei** [comment]
Going with Module for now sounds good. We should bring this up on the mailing list as soon as possible and talk about this on Friday's design review.

**Comment by eaplatanios** [comment]
@rxwei I'll being it up later today. Is the toolchain updated in the CI? Because I think the test failures are due to an older toolchain.

**Comment by rxwei** [comment]
IIRC nightly toolchains are tracking the tensorflow-0.4 branch right now. @bgogul

**Comment by eaplatanios** [comment]
I see. That would mean they do not include the fix from apple/swift#25974 which is what causing the CI failures I believe.

**Comment by eaplatanios** [comment]
@rxwei I ran tests about an hour ago. Unless the toolchain has been updated in the meantime, the tests won't pass.

**Comment by rxwei** [comment]
A new toolchain was produced at ~10:45ish.

**Comment by eaplatanios** [comment]
Coincidentally, [that's exactly when I ran them](https://source.cloud.google.com/results/invocations/a66d2554-cbfa-49bf-9dc5-b0b164a4ee1a/targets/s4tf%2Fapis%2Fubuntu%2Fpresubmit/log) so it may succeed now. :)

**Comment by eaplatanios** [comment]
This passes locally so I'm not sure what's wrong here. Are you sure the toolchain is the latest one from the `tensorflow` branch? If it is from the `tensorflow-0.4` branch are you sure that the conditional differentiable PR has been merged there?

**Comment by JacopoMangiavacchi** [comment]
I see we already have an initializer for using a pretrained embeddings table but it is not clear for me how we can froze also the embedding table itself. Keras use to have a **trainable** parameter in the constructor.  Don't you think we should introduce that parameter too to this initializer?  If value is true we should do not differentiate the weights table too isn't it?

**Comment by eaplatanios** [comment]
@JacopoMangiavacchi there is separate ongoing work on using property wrappers to allow freezing trainable parameters more generally for layers. For that reason, for now we follow the same convention we've been using for layers all along.

**Comment by rxwei** [comment]
@eaplatanios Could you update the PR description to accurately describe the changes?

**Comment by eaplatanios** [comment]
@rxwei done.

**Comment by rxwei** [comment]
Sorry, I meant PR description (not just the title). If you could list out the various changes, that'd be great. We want to include that in the commit log.

**Comment by eaplatanios** [comment]
Sorry my bad. I just updated the description.
